File: longest_palindrome.cpp
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach
// Idea is to add the characters to the set initially and when we encounter
// them again erase those characters and increment count by 2. At the end if the size of 
// set is > 0 then we add 1 to the count. This gives the length of longest palindrome.

#include<string>
#include<unordered_set>

class Solution {
public:
    int longestPalindrome(std::string s) {
        int count = 0;
        if(s.length() == 0)
        {
            return count;
        }

        // to track characters.
        std::unordered_set<char> set;
        for(int i = 0; i< s.length(); i++)
        {
            char ch = s[i];
            if(set.find(ch) == set.end())
            {
                // When set doesnt have that char
                set.insert(ch);
            }
            else
            {
                // when set already has that char
                set.erase(ch);
                count = count+2;
            }
        }
        
        if(set.size() > 0)
        {
            count++;
        }
        return count;
    }
};

File: longestpalindrome.go
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
in the map if we already have one existance and encounter again then remove from map and increase count by 2
if its not present in map then just set to true
at last if any one of the element is true, increase count by 1
*/
package main

import "fmt"

func longestPalindrome(s string) int {
	count := 0
	arr := map[rune]bool{}
	for _, v := range s {
		if arr[v] == false {
			arr[v] = true
		} else {
			arr[v] = false
			count = count + 2
		}
	}
	for _, v := range arr {
		if v == true {
			count++
			return count
		}
	}
	return count
}

func MainLongestPalindrome() {
	fmt.Println(longestPalindrome("abccccdd")) //expected 7
}
